src/c/tb_client/signal.zig

```diff
@@ -274,8 +274,12 @@ pub const Signal = struct {
             .Acquire,
             .Acquire,
         ) orelse {
-            (self.on_signal_fn)(self);
-            return self.wait();
+            const pending = (self.on_signal_fn)(self);
+            self.wait();
+
+            // Run again on the next tick if there are still packets to process.
+            if (pending) self.notify();
```

Comment:
I think the `notify()` should be done before the `wait()` to go through the yield path of `io.timeout` (observing `.notified` at the CAS) instead of going through `io.recv` (which results in a syscall on non-io_uring platforms).

Reply:
Calling `notify()` before `wait()` hangs too.
It's not intuitive for me, but this time it printed this message after a few seconds running:

```
info(message_bus): message queue for peer .tigerbeetle.lib.tigerbeetle.src.message_bus.union:375:19{ .replica = 0 } full, dropping ping message
```

Strange enough, printing a debug message to stderr causes that same code to work most of the time, but it still hangs eventually.

```diff
+      std.log.err("timeout ...", .{});
        self.io.timeout(
            *Signal,
            self,
            on_timeout,
            &self.completion,
            0, // zero-timeout functions as a yield
        );
```

Reply:
After some thought, should the notify() happen in `on_complete`? Thats where the available_messages is incremented again and is where a retry should occur

Reply:
Yes, I've thought of that, but it would be called multiple times needless, if there are many requests being completed on the same tick.

Reply:
> if there are many requests being completed on the same tick.

Since a client only has one request in-flight at a time, is that possible?

Reply:
Let me try ti call notify only when available_messages == 1

Reply:
notify() is cheap to call as it only does a syscall if the signal is waiting. If notify() is happning in the ClientThread code, the Signal is running the callback and not waiting.

Reply:
Much better now, thank you!